mlir/lib/Catalyst/Transforms/catalyst_to_llvm.cpp

```diff
@@ -232,6 +263,174 @@ struct PrintOpPattern : public OpConversionPattern<PrintOp> {
+    Value data = rewriter.create<LLVM::GEPOp>(loc, ptr, memrefType.getElementType(),
+                                              desc.alignedPtr(rewriter, loc), c0);
+    memref = rewriter.create<LLVM::InsertValueOp>(loc, memref, data, 1);
+
+    // Dtype
+    memref = rewriter.create<LLVM::InsertValueOp>(loc, memref, dtype, 2);
+
+    return memref;
+}
+
+struct CustomCallOpPattern : public OpConversionPattern<CustomCallOp> {
+    using OpConversionPattern::OpConversionPattern;
+
+    LogicalResult matchAndRewrite(CustomCallOp op, CustomCallOpAdaptor adaptor,
+                                  ConversionPatternRewriter &rewriter) const override
+    {
+        MLIRContext *ctx = op.getContext();
+        Location loc = op.getLoc();
+        // Create function
+        Type ptr = LLVM::LLVMPointerType::get(ctx);
+
+        Type voidType = LLVM::LLVMVoidType::get(ctx);
+        auto type = LLVM::LLVMFunctionType::get(voidType, {/*args=*/ptr, /*rets=*/ptr});
+        auto point = rewriter.saveInsertionPoint();
+        ModuleOp mod = op->getParentOfType<ModuleOp>();
+        rewriter.setInsertionPointToStart(mod.getBody());
+        LLVM::LLVMFuncOp customCallFnOp =
+            rewriter.create<LLVM::LLVMFuncOp>(loc, op.getCallTargetName(), type);
+        customCallFnOp.setPrivate();
+        rewriter.restoreInsertionPoint(point);
+        // Setup args and res
+        int32_t numberArg = op.getNumberOriginalArgAttr()[0];
+        SmallVector<Value> operands = op.getOperands();
+        SmallVector<Value> args = {operands.begin(), operands.begin() + numberArg};
+        SmallVector<Value> res = {operands.begin() + numberArg, operands.end()};
+
+        SmallVector<Value> operandsConverted = adaptor.getOperands();
+        SmallVector<Value> argsConverted = {operandsConverted.begin(),
+                                            operandsConverted.begin() + numberArg};
+        SmallVector<Value> resConverted = {operandsConverted.begin() + numberArg,
+                                           operandsConverted.end()};
+
+        // Encode args
+        Value c1 = rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI64IntegerAttr(1));
+        SmallVector<LLVM::AllocaOp> encodedArgs;
+        for (auto tuple : llvm::zip(args, argsConverted)) {
+            auto memref_type = std::get<0>(tuple).getType().cast<MemRefType>();
+            Type llvmMemrefType = std::get<1>(tuple).getType();
+            auto encodedArg =
+                EncodeDataMemRef(loc, rewriter, memref_type, llvmMemrefType, std::get<1>(tuple));
+            LLVM::AllocaOp alloca =
+                rewriter.create<LLVM::AllocaOp>(loc, ptr, encodedArg.getType(), c1, 0);
+            // Use volatile store to suppress expensive LLVM optimizations.
+            rewriter.create<LLVM::StoreOp>(loc, encodedArg, alloca, /*alignment=*/0,
+                                           /*isVolatile=*/true);
```

Comment:
Out of curiosity, what are the expensive optimization, and how does volatile affect them? :)

Reply:
From LLVM doc:

```
There are two arguments to the store instruction: a value to store and an address at which to store it. The type of the <pointer> operand must be a pointer to the [first class](https://llvm.org/docs/LangRef.html#t-firstclass) type of the <value> operand. If the store is marked as volatile, then the optimizer is not allowed to modify the number or order of execution of this store with other [volatile operations](https://llvm.org/docs/LangRef.html#volatile). Only values of [first class](https://llvm.org/docs/LangRef.html#t-firstclass) types of known size (i.e. not containing an [opaque structural type](https://llvm.org/docs/LangRef.html#t-opaque)) can be stored.
```

Some optimizer passes tries to optimize the number of store and also the order of execution. I was thinking it can be useful for a high number of store. But for our purpose it might be overkill, therefore I removed it.

Reply:
Interesting, I was just curious :)